### PR TITLE
Refactor numeric_cast

### PR DESCRIPTION
**Pull request by reuk**

### CODE REVIEW

File: src/util/mp_arith.h [hunk 1]
```diff
@@ -62,82 +64,93 @@ unsigned integer2unsigned(const mp_integer &);
...
+  template <typename U = T,
+            typename std::enable_if<!std::is_signed<U>::value, int>::type = 0>
+  static auto get_val(const mp_integer &mpi) -> decltype(mpi.to_ulong())
+  {
+    return mpi.to_ulong();
+  }
+
+  optionalt<T> operator()(const mp_integer &mpi) const
   {
+    constexpr const auto precondition =
```
Comment (smowton) [inline src/util/mp_arith.h]: Surely this will be a static error on VS2013 which doesn't have constexprs in the standard library

Comment (reuk) [reply src/util/mp_arith.h]: Hmm, potentially yeah. You could put an ifdef around the constexpr I suppose, or we might have a macro that conditionally expands to constexpr in util somewhere.

File: src/util/mp_arith.h [hunk 2]
```diff
@@ -62,82 +64,93 @@ unsigned integer2unsigned(const mp_integer &);
...
 struct numeric_castt<T,
-                     mp_integer,
-                     typename std::enable_if<std::is_integral<T>::value &&
-                                             std::is_signed<T>::value>::type>
+                     typename std::enable_if<std::is_integral<T>::value>::type>
 {
-  static optionalt<T> numeric_cast(const mp_integer &mpi)
+  template <typename U = T,
+            typename std::enable_if<std::is_signed<U>::value, int>::type = 0>
+  static auto get_val(const mp_integer &mpi) -> decltype(mpi.to_long())
```
Comment (reuk) [inline src/util/mp_arith.h:114]: These should both be private

**Review by romainbrenguier — COMMENTED** [review]
Thanks for the proposal. What I don't find nice is that we pull a lot of things from `arith_tools` to `mp_arith` which I think we want to keep seperate. Could we instead define these `numeric_cast` things in a new file?

File: src/util/mp_arith.h [hunk 3]
```diff
@@ -62,82 +64,93 @@ unsigned integer2unsigned(const mp_integer &);
 
 const mp_integer mp_zero=string2integer("0");
 
+class exprt;
```
Comment (romainbrenguier) [inline src/util/mp_arith.h:67]: duplicate of line 23?

**Comment by reuk** [comment]
The problem is that I think all the numeric cast overloads need to be defined in one go. So as long as all the numeric_cast definitions are moved, that will work, but moving just a few of them won't

**Comment by reuk** [comment]
I'm not sure what to do with this branch. I don't plan to work on it any more, as all the suggested changes are very minor. I'd suggest that you cherry-pick any changes you like (if any) and close it once you've extracted anything useful.

**Comment by romainbrenguier** [comment]
@reuk 
> I'm not sure what to do with this branch. I don't plan to work on it any more, as all the suggested changes are very minor. I'd suggest that you cherry-pick any changes you like (if any) and close it once you've extracted anything useful.

Yes no problem, I already integrated some of the changes in my branch. Just leave this open in case there are other changes I could use.

**Pull request closed without merging**